lib/memoryprofiler/src/trait.rs

```diff
@@ -0,0 +1,169 @@
+pub trait MemoryUsage {
+    fn size_of(&self) -> usize;
+}
+
+/// Primitive types
+impl MemoryUsage for bool {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<bool>()
+    }
+}
+impl MemoryUsage for char {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<char>()
+    }
+}
+impl MemoryUsage for f32 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<f32>()
+    }
+}
+impl MemoryUsage for f64 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<f64>()
+    }
+}
+impl MemoryUsage for i8 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i8>()
+    }
+}
+impl MemoryUsage for i16 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i16>()
+    }
+}
+impl MemoryUsage for i32 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i32>()
+    }
+}
+impl MemoryUsage for i64 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i64>()
+    }
+}
+impl MemoryUsage for isize {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<isize>()
+    }
+}
+impl MemoryUsage for u8 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u8>()
+    }
+}
+impl MemoryUsage for u16 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u16>()
+    }
+}
+impl MemoryUsage for u32 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u32>()
+    }
+}
+impl MemoryUsage for u64 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u64>()
+    }
+}
+impl MemoryUsage for usize {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<usize>()
+    }
+}
+
+// pointer
+// Pointers aren't necessarily safe to dereference, even if they're nonnull.
+
+// references
+impl<T: MemoryUsage> MemoryUsage for &T {
+    fn size_of(&self) -> usize {
+        (**self).size_of()
+    }
+}
+impl<T: MemoryUsage> MemoryUsage for &mut T {
+    fn size_of(&self) -> usize {
+        (**self).size_of()
+    }
+}
+
+// slices
+impl<T: MemoryUsage> MemoryUsage for [T] {
+    fn size_of(&self) -> usize {
+        self.iter().map(MemoryUsage::size_of).sum()
+    }
+}
+
+// arrays
+impl<T: MemoryUsage, const N: usize> MemoryUsage for [T; N] {
+    fn size_of(&self) -> usize {
+        self.iter().map(MemoryUsage::size_of).sum()
+    }
+}
+
+// strs
+impl MemoryUsage for str {
+    fn size_of(&self) -> usize {
+        self.as_bytes().size_of()
+    }
+}
+
+// TODO: tuples
+
+/// Standard library types
+
+// TODO: Arc
+
+//impl<T: MemoryUsage> MemoryUsage for Box<T> {
+//}
+
+// Cell
+
+// Is a Pin always dereferenceable?
+//impl<T: MemoryUsage> MemoryUsage for Pin<T> {
+//}
+
+// TODO: Mutex
+
+// TODO: NonNull might be possible when '*const T' is MemoryUsage.
+
+impl<T: MemoryUsage> MemoryUsage for Option<T> {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<Option<()>>() + self.iter().map(MemoryUsage::size_of).sum::<usize>()
```

Comment:
I don't know if this is correct, Option does optimizations depending on what it's holding so for example

`NonNull<T>` and `Option<NonNull<T>>` are the same size

Reply:
Alright lets break it down: we know in the `None` case it's always just the size of the stack space taken up so `std::mem::size_of::<Option<T>>()` is all we need.

In the `Some` case we need to know the memory usage of the inner type because of heap usage and then `std::mem::size_of::<Option<T>>()`: something I've been thinking about with this trait is having another method and trying to make a distinction between the static size allocated and the dynamic size.

We might actually need that to make that work. A neat thing we can do for the static one is basically just have it be `std::mem::size_of::<T>()` where it takes the argument by generic type.

I don't think we can do `MemoryUsage: Sized` and rely on `std::mem::size_of` for this though because that would prevent this trait from being derived by other traits that are then used as trait objects: the desired behavior we want for traits with no real stack size is to say that they don't have one (this applies for unsized types like `[u8]` too)


A bit stream of consciousness but ping me if you want to discuss anything about this today

Reply:
I think `std::mem::size_of::<Option<T>>() + self.iter().map(MemoryUsage::size_of).sum::<usize>()` gets exactly what we want here? The size of the Option including NonNull optimization if applicable, then visit the member if it's holding one?